File: oblig2/Legemiddel.java
/*
  Legemiddel er en abstract class siden det alle virkelige legemidler tilhører
  en spesifik type. 
*/

abstract class Legemiddel {

    // Variabler opprettes. Count brukes for å telle alle gangene
    // Legemiddel opprettes (er static så blir sett felles for alle
    // objekter av typen Legemiddel), og id er et unikt tall for
    // alle objekter av typen Legemiddel.
    private static int count = 0;
    protected int id;
    protected String navn;
    protected double pris;
    protected double virkestoff;

    // Konstruktør
    Legemiddel(String navnet, double prisen, double virkestoffet) {
	navn = navnet;
	pris = prisen;
	virkestoff = virkestoffet;
	id = count++;
    }

    // Metoder
    // getters; returnerer variabler som klassen tar vare på
    public int hentId() {
	return id;
    }

    public String hentNavn() {
	return navn;
    }

    public double hentPris() {
	return pris;
    }

    public double hentVirkestoff() {
	return virkestoff;
    }

    // setters; setter variabler klassen tar vare på
    public void settNyPris(double nyPris) {
	pris = nyPris;
    }

    // Utskrifter
    @Override
    public String toString() {
	return "ID: " + hentId() + "\t- Navn: " + hentNavn() + "\t- Pris: " + hentPris() + "\t- Virkestoff: " + hentVirkestoff();
    }
    
}

File: oblig2/cpp/Legemiddel.cpp
#include "Legemiddel.h"
#include <utility>

/* 
   Legemiddel
*/
Legemiddel::Legemiddel(const char* navn, double pris, const double virkestoff)
  : m_navn(std::move(navn)), m_pris(std::move(pris)), m_virkestoff(std::move(virkestoff)),
    m_id(std::move(count++))
{}
int Legemiddel::count = 0;
const int Legemiddel::hentId() const { return m_id; }
const char* Legemiddel::hentNavn() const { return m_navn; }
double Legemiddel::hentPris() const { return m_pris; }
const double Legemiddel::hentVirkestoff() const { return m_virkestoff; }

std::ostream& operator<<(std::ostream& out, const Legemiddel& L) {
  return out << "ID: " << L.hentId() << "\t- Navn: " << L.hentNavn() <<
    "\t- Pris: " << L.hentPris() << "\t- Virkestoff: " << L.hentVirkestoff();
}

/*
  Narkotisk
*/
Narkotisk::Narkotisk(const char* navn, double pris, const double virkestoff, const int styrke)
  : Legemiddel{ navn, pris, virkestoff }, m_styrke(std::move(styrke))
{}
const int Narkotisk::hentStyrke() const { return m_styrke; }

std::ostream& operator<<(std::ostream& out, const Narkotisk& N) {
  return out << "Narkotisk\t" << (Legemiddel) N << "\t- Styrke: " << N.hentStyrke();
}

/*
  Vanedannende
*/
Vanedannende::Vanedannende(const char* navn, double pris, const double virkestoff,
			   const int vanedannendeStyrke)
  : Legemiddel{ navn, pris, virkestoff }, m_vanedannendeStyrke(std::move(vanedannendeStyrke))
{}
const int Vanedannende::hentVaneStyrke() const { return m_vanedannendeStyrke; }

std::ostream& operator<<(std::ostream& out, const Vanedannende& V) {
  return out << "Vanedannende\t" << (Legemiddel) V << "\t- Styrke: " << V.hentVaneStyrke();
}

/*
  Vanlig
*/
Vanlig::Vanlig(const char* navn, double pris, const double virkestoff)
  : Legemiddel{ navn, pris, virkestoff }
{}

std::ostream& operator<<(std::ostream& out, const Vanlig& V) {
  return out <<"Vanlig\t\t" << (Legemiddel) V;
}
